File: Tree/Hard/124. binary-tree-maximum-path-sum/binary-tree-maximum-path-sum_GlobalVariable.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Version 1: Global Variable
public class Solution {
    // initiate the maxSum as the MinValue of int
    private int maxSum = Int32.MinValue;
    
    public int MaxPathSum(TreeNode root) {
        // recursive
        // TC: O(n); SC: O(h)
        if (root == null) {
            return 0;
        }
        
        GetMaxOnOneSide(root);
        
        return maxSum;
    }
    
    private int GetMaxOnOneSide(TreeNode node) {
        if (node == null) {
            return 0;
        }
        
        // value < 0 would not contribute to the max path
        var left = Math.Max(GetMaxOnOneSide(node.left), 0);
        var right = Math.Max(GetMaxOnOneSide(node.right), 0);
        
        // start a new path with current node as the root (highest node)
        int priceNewPath = node.val + left + right;
        maxSum = Math.Max(maxSum, priceNewPath);
        
        // for recursion: return the max gain if continue the same path from bottom to top
        return node.val + Math.Max(left, right);
    }
}

File: Tree/Hard/124. binary-tree-maximum-path-sum/binary-tree-maximum-path-sum_GlobalVariable.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Version 1: Global Variable
class Solution:
    # initiate maxSum as the MinValue of int
    def __init__(self):
        self.maxSum = float('-inf')
    
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.getMaxOnOneSide(root)
        return self.maxSum
    
    def getMaxOnOneSide(self, node):            
        if not node:
            return 0

        # nonlocal maxSum
        # value < 0 would not contribute to the max path
        left = max(self.getMaxOnOneSide(node.left), 0)
        right = max(self.getMaxOnOneSide(node.right), 0)

        # start a new path with current node as the root (highest node)
        priceNewPath = node.val + left + right;
        self.maxSum = max(self.maxSum, priceNewPath)

        # for recursion: return the max gain if continue the same path from bottom to top
        return node.val + max(left, right)